kinfer/export/pytorch.py

```diff
@@ -0,0 +1,204 @@
+"""PyTorch model export utilities."""
+
+import inspect
+from dataclasses import fields, is_dataclass
+from io import BytesIO
+from typing import Any, Dict, List, Optional, Tuple, Union
+
+import onnx
+import onnxruntime as ort
+import torch
+from torch import nn
+
+
+def get_model_info(model: nn.Module) -> Dict[str, Any]:
+    """Extract model information including input parameters and their types.
+
+    Args:
+        model: PyTorch model to analyze
+
+    Returns:
+        Dictionary containing model information
+    """
+    # Get model's forward method signature
+    signature = inspect.signature(model.forward)
+
+    # Extract parameter information
+    params_info = {}
+    for name, param in signature.parameters.items():
+        if name == 'self':
+            continue
+        params_info[name] = {
+            'annotation': str(param.annotation),
+            'default': None if param.default is param.empty else str(param.default)
+        }
+
+    return {
+        'input_params': params_info,
+        'num_parameters': sum(p.numel() for p in model.parameters()),
+    }
+
+def add_metadata_to_onnx(
+    model_proto: onnx.ModelProto,
+    metadata: Dict[str, Any],
+    config: Optional[object] = None
+) -> onnx.ModelProto:
+    """Add metadata to ONNX model.
+
+    Args:
+        model_proto: ONNX model prototype
+        metadata: Dictionary of metadata to add
+        config: Optional configuration dataclass to add to metadata
+
+    Returns:
+        ONNX model with added metadata
+    """
+    # Add model metadata
+    for key, value in metadata.items():
+        meta = model_proto.metadata_props.add()
+        meta.key = key
+        meta.value = str(value)
+
+    # Add configuration if provided
+    if config is not None and is_dataclass(config):
+        for field in fields(config):
+            value = getattr(config, field.name)
+            meta = model_proto.metadata_props.add()
+            meta.key = field.name
+            meta.value = str(value)
+
+    return model_proto
+
+def infer_input_shapes(model: nn.Module) -> Union[torch.Size, List[torch.Size]]:
+    """Infer input shapes from model architecture.
+
+    Args:
+        model: PyTorch model to analyze
+
+    Returns:
+        Single input shape or list of input shapes
+    """
+    # Try to find first layer
+    first_layer = None
+
+    # Check if model is Sequential
+    if isinstance(model, nn.Sequential):
+        first_layer = model[0]
+    # Check if model has named modules
+    else:
+        for module in model.modules():
+            if isinstance(module, (nn.Linear, nn.Conv1d, nn.Conv2d, nn.Conv3d)):
+                first_layer = module
+                break
```

Comment:
this doesn't seem like it will work reliably?

Reply:
I think we probably want to explicitly pass input_tensors to the export function, but I also wanted to copy over the inference I had for ppo